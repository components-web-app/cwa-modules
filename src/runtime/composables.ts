import { useNuxtApp } from '#imports'

export const useCwa = () => useNuxtApp().$cwa
