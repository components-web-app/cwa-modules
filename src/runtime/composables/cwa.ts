import { useNuxtApp } from '#imports'
import type Cwa from '#cwa/runtime/cwa'

export const useCwa = () => useNuxtApp().$cwa as Cwa
