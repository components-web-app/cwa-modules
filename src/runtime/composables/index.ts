import { useNuxtApp } from '#imports'

export const useCwa = () => useNuxtApp().$cwa
export const useCwaResources = () => useNuxtApp().$cwa.resources
export const useCwaResourcesManager = () => useNuxtApp().$cwa.resourcesManager
export const useCwaAuth = () => useNuxtApp().$cwa.auth
export const useCwaForms = () => useNuxtApp().$cwa.forms
