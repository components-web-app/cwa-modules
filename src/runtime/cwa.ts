import { NuxtApp } from '#app/nuxt'
import { CwaModuleOptions } from '../module'
import { Storage } from './storage/storage'
import { Fetcher } from './api/fetcher/fetcher'
import Mercure from './api/mercure'
import ApiDocumentation from './api/api-documentation'

export default class Cwa {
  private readonly nuxtApp: NuxtApp
  private readonly options: CwaModuleOptions
  private readonly storage: Storage
  public readonly mercure: Mercure
  public readonly apiDocumentation: ApiDocumentation
  public readonly fetcher: Fetcher

  constructor (nuxtApp: NuxtApp, options: CwaModuleOptions) {
    const apiUrl = options.apiUrlBrowser || options.apiUrl || 'https://api-url-not-set.com'

    this.nuxtApp = nuxtApp
    this.options = options
    this.storage = new Storage(this.options.storeName)
    const currentRoute = nuxtApp._route
    this.mercure = new Mercure(this.storage.stores.mercure)
    this.apiDocumentation = new ApiDocumentation(this.storage.stores.apiDocumentation)
    this.fetcher = new Fetcher(apiUrl, this.storage.stores.fetcher, this.storage.stores.resources, currentRoute, this.mercure, this.apiDocumentation)
  }

  public get stores () {
    return this.storage.stores
  }

  public initMercure () {
    this.mercure.init()
  }
}
