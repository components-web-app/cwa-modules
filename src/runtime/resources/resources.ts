import { computed } from 'vue'
import type { ComputedRef } from 'vue'
import { ResourcesStore } from '../storage/stores/resources/resources-store'
import { CwaResourceApiStatuses, NEW_RESOURCE_IRI } from '../storage/stores/resources/state'
import type { CwaCurrentResourceInterface } from '../storage/stores/resources/state'
import { FetcherStore } from '../storage/stores/fetcher/fetcher-store'
import type { FetchStatus } from '../storage/stores/fetcher/state'
import {
  CwaResourceTypes,
  getResourceTypeFromIri
} from './resource-utils'
import type ResourceStackManager from '#cwa/runtime/admin/resource-stack-manager'

interface PageLoadStatus {
  resources: (string|undefined)[]
  total: number
  complete: number
  percent: number
}

export class Resources {
  // eslint-disable-next-line no-useless-constructor
  constructor (private readonly resourcesStoreDefinition: ResourcesStore, private readonly fetcherStoreDefinition: FetcherStore, private readonly resourceManager: ResourceStackManager) {
  }

  public get currentIds () {
    return this.resourcesStore.current.currentIds
  }

  public isIriPublishableEquivalent (oldIri: string, newIri: string) {
    return this.resourcesStore.isIriPublishableEquivalent(oldIri, newIri)
  }

  public findAllPublishableIris (iri: string) {
    return this.resourcesStore.findAllPublishableIris(iri)
  }

  public getResource (id: string) {
    return computed(() => {
      return this.resourcesStore.getResource(id)
    })
  }

  public get newResource () {
    return computed(() => this.resourcesStore.getResource(NEW_RESOURCE_IRI))
  }

  public getComponentGroupByReference (reference: string) {
    const componentGroups = this.resourcesStore.resourcesByType[CwaResourceTypes.COMPONENT_GROUP]
    return componentGroups.find((componentGroupResource) => {
      return componentGroupResource.data?.reference === reference
    })
  }

  public get currentResources () {
    return this.resourcesStore.current.currentIds.reduce((obj, id: string) => {
      const idResource = this.getResource(id).value
      if (idResource) {
        obj[id] = idResource
      }
      return obj
    }, {} as {
      [key: string]: CwaCurrentResourceInterface
    })
  }

  private get displayFetchStatus () {
    const fetchingToken = this.fetcherStore.primaryFetch.fetchingToken
    // if the page is fetched in a primary fetching token in progress we start showing that page load progress
    if (fetchingToken) {
      const fetchingStatus = this.fetcherStore.fetches[fetchingToken]
      if (fetchingStatus) {
        const pageIri = this.getPageIriByFetchStatus(fetchingStatus)
        if (pageIri && this.resourcesStore.current.currentIds.includes(pageIri)) {
          const pageResource = this.getResource(pageIri).value
          if (pageResource?.data && pageResource.apiState.status === CwaResourceApiStatuses.SUCCESS) {
            return fetchingStatus
          }
        }
      }
    }
    return this.fetcherStore.resolvedSuccessFetchStatus
  }

  private get pageLoadResources () {
    const token = this.fetcherStore.primaryFetch.fetchingToken
    if (!token) {
      return
    }
    const fetchStatus = this.fetcherStore.fetches[token]
    if (!fetchStatus) {
      return
    }
    const type = this.getFetchStatusType(fetchStatus)
    if (!type) {
      return
    }

    const resources: (string|undefined)[] = [
      this.getLayoutIriByFetchStatus(fetchStatus),
      this.getPageIriByFetchStatus(fetchStatus)
    ]

    if (type === CwaResourceTypes.ROUTE || type === CwaResourceTypes.PAGE_DATA) {
      resources.push(fetchStatus.path)
    }

    if (type === CwaResourceTypes.ROUTE) {
      const routeResource = this.getResource(fetchStatus.path).value
      if (routeResource) {
        const pageDataIri = routeResource.data?.value?.pageData
        if (pageDataIri) {
          resources.push(pageDataIri)
        }
      }
    }

    return resources
  }

  public get pageLoadProgress (): ComputedRef<PageLoadStatus> {
    return computed<PageLoadStatus>(() => {
      const pageLoadResources = this.pageLoadResources
      if (!pageLoadResources) {
        return {
          resources: [],
          total: 0,
          complete: 0,
          percent: 100
        }
      }

      const total = pageLoadResources.length
      let complete = 0
      for (const resourceIri of pageLoadResources) {
        if (!resourceIri) {
          continue
        }
        const resource = this.getResource(resourceIri).value
        if (resource?.apiState.status !== CwaResourceApiStatuses.IN_PROGRESS) {
          complete++
        }
      }

      const percent = complete ? Math.round((complete / total) * 100) : 0

      return {
        resources: pageLoadResources,
        total,
        complete,
        percent
      }
    })
  }

  public isPageDataResource (iri: Ref<string|undefined>) {
    return computed(() => {
      if (!this.pageData?.value?.data || !iri.value) {
        return false
      }
      if (getResourceTypeFromIri(iri.value) !== CwaResourceTypes.COMPONENT) {
        return false
      }
      return Object.values(this.pageData.value.data).includes(iri.value)
    })
  }

  private getFetchStatusType (fetchStatus?: FetchStatus): undefined|string {
    if (!fetchStatus) {
      return
    }
    const type = getResourceTypeFromIri(fetchStatus.path)
    if (!type) {
      return
    }
    return type
  }

  private getLayoutIriByFetchStatus (fetchStatus?: FetchStatus): string|undefined {
    const pageIri = this.getPageIriByFetchStatus(fetchStatus)
    if (!pageIri) {
      return
    }
    const pageResource = this.getResource(pageIri).value
    return pageResource?.data?.layout
  }

  private getPageIriByFetchStatus (fetchStatus?: FetchStatus): string|undefined {
    const type = this.getFetchStatusType(fetchStatus)
    if (!type) {
      return
    }
    if (!fetchStatus?.path) {
      return
    }

    if (type === CwaResourceTypes.PAGE) {
      return fetchStatus.path
    }
    const successResource = this.getResource(fetchStatus.path).value
    if (!successResource) {
      return
    }
    switch (type) {
      case CwaResourceTypes.PAGE_DATA: {
        return successResource.data?.page
      }
      case CwaResourceTypes.ROUTE: {
        const pageData = successResource.data?.pageData
        if (pageData) {
          const pageDataResource = this.getResource(pageData).value
          return pageDataResource?.data?.page
        }
        return successResource.data?.page
      }
    }
  }

  public get pageDataIri () {
    return computed(() => {
      const fetchStatus = this.displayFetchStatus
      const type = this.getFetchStatusType(fetchStatus)
      if (!type) {
        return
      }
      if (!fetchStatus?.path) {
        return
      }
      if (type === CwaResourceTypes.PAGE_DATA) {
        return fetchStatus.path
      }
      const successResource = this.getResource(fetchStatus.path).value
      return type === CwaResourceTypes.ROUTE ? successResource?.data?.pageData : undefined
    })
  }

  public get pageData () {
    if (!this.pageDataIri.value) {
      return
    }
    return this.getResource(this.pageDataIri.value)
  }

  public get pageIri (): ComputedRef<string|undefined> {
    return computed(() => this.getPageIriByFetchStatus(this.displayFetchStatus))
  }

  public get page () {
    if (!this.pageIri.value) {
      return
    }
    return this.getResource(this.pageIri.value)
  }

  public get layoutIri (): ComputedRef<string|undefined> {
    return computed(() => this.getLayoutIriByFetchStatus(this.displayFetchStatus))
  }

  public get layout (): ComputedRef<CwaCurrentResourceInterface|undefined> {
    return computed(() => {
      const layoutIri = this.layoutIri.value
      if (!layoutIri) {
        return
      }
      return this.getResource(layoutIri).value
    })
  }

  public findPublishedComponentIri (iri: string) {
    return computed(() => {
      return this.resourcesStore.findPublishedComponentIri(iri)
    })
  }

  public findDraftComponentIri (iri: string) {
    return computed(() => {
      return this.resourcesStore.findDraftComponentIri(iri)
    })
  }

  public get isLoading (): ComputedRef<boolean> {
    return computed(() => {
      return !this.fetcherStore.fetchesResolved || !!this.resourceLoadStatus.pending
    })
  }

  public get resourceLoadStatus () {
    return this.resourcesStore.resourceLoadStatus
  }

  private get fetcherStore () {
    return this.fetcherStoreDefinition.useStore()
  }

  private get resourcesStore () {
    return this.resourcesStoreDefinition.useStore()
  }

  public get usesPageTemplate () {
    return computed(() => !!this.page?.value?.data?.isTemplate)
  }

  public get isDataPage () {
    return computed(() => {
      return this.usesPageTemplate.value && !!this.pageDataIri.value
    })
  }

  public get isDynamicPage () {
    return computed(() => {
      return this.usesPageTemplate.value && !this.pageDataIri.value
    })
  }

  public get hasNewResources () {
    return this.resourcesStore.hasNewResources
  }
}
