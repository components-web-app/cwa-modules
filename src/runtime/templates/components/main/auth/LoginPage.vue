<template>
  <AuthTemplate
    :submitting="submitting"
    :submit-button-text="submitButtonText"
    :error="error"
    @submit="$emit('submit')"
  >
    <template v-if="credentials">
      <InputField
        v-model="credentials.username"
        label="Username or Email"
        name="username"
        type="text"
        autocomplete="username"
        :required="true"
      />
      <InputField
        v-model="credentials.password"
        label="Password"
        name="password"
        type="password"
        autocomplete="current-password"
        :required="true"
      />
    </template>
    <slot>
      <div class="cwa-flex cwa-items-center cwa-justify-between">
        <AuthPageLink link-to="/" link-text="< Back to home" />
        <AuthPageLink link-to="/forgot-password" link-text="Forgot your password?" />
      </div>
    </slot>
  </AuthTemplate>
</template>

<script setup lang="ts">
import { computed, defineAsyncComponent } from 'vue'
import AuthTemplate from './_parts/AuthTemplate.vue'
import InputField from './_parts/InputField.vue'

const emit = defineEmits(['submit', 'update:modelValue'])

const props = defineProps<{
  submitButtonText: string,
  submitting: boolean,
  modelValue: {
    username: string,
    password: string
  }
  error?: string
}>()

const AuthPageLink = defineAsyncComponent(() =>
  import('./_parts/AuthPageLink.vue')
)

const credentials = computed({
  get () {
    return props.modelValue
  },
  set (newValue) {
    emit('update:modelValue', newValue)
  }
})
</script>
