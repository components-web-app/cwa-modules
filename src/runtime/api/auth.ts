import { FetchError } from 'ofetch'
import { computed, ref } from 'vue'
import type { ComputedRef, Ref } from 'vue'
import { useRoute } from '#app'
import type { CookieRef } from '#app'
import { AuthStore } from '../storage/stores/auth/auth-store'
import { CwaUserRoles } from '../storage/stores/auth/state'
import { ResourcesStore } from '../storage/stores/resources/resources-store'
import { FetcherStore } from '../storage/stores/fetcher/fetcher-store'
import CwaFetch from './fetcher/cwa-fetch'
import Mercure from './mercure'
import Fetcher from './fetcher/fetcher'
import Admin from '#cwa/runtime/admin/admin'

interface Credentials {
  username: string
  password: string
}

export enum CwaAuthStatus {
  SIGNED_OUT = 0,
  LOADING = 1,
  SIGNED_IN = 2
}

interface ResetPasswordEvent {
  username: string
  token: string
  passwords: {
    first: string
    second: string
  }
}

export default class Auth {
  private loading: Ref<boolean>
  private hasCheckedMeEndpointForInit = false

  public constructor (
    private readonly cwaFetch: CwaFetch,
    private readonly mercure: Mercure,
    private readonly fetcher: Fetcher,
    private readonly admin: Admin,
    private readonly authStoreDefinition: AuthStore,
    private readonly resourcesStoreDefinition: ResourcesStore,
    private readonly fetcherStoreDefinition: FetcherStore,
    private readonly authCookie: CookieRef<string | null>
  ) {
    this.loading = ref(false)
  }

  public async signIn (credentials: Credentials) {
    const result = await this.loginRequest(credentials)
    if (result instanceof FetchError) {
      return result
    }
    this.mercure.init(true)
    return this.refreshUser()
  }

  public async forgotPassword (username: string) {
    try {
      return await this.cwaFetch.fetch(`/password/reset/request/${encodeURIComponent(
        username
      )}`)
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    }
  }

  public async resetPassword (event: ResetPasswordEvent) {
    try {
      return await this.cwaFetch.fetch('/component/forms/password_reset/submit', {
        method: 'PATCH',
        body: {
          password_update: {
            username: event.username,
            plainNewPasswordConfirmationToken: event.token,
            plainPassword: event.passwords
          }
        }
      })
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    }
  }

  public async signOut () {
    this.loading.value = true
    try {
      const result = await this.cwaFetch.fetch('/logout')
      await this.clearSession()
      return result
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    } finally {
      this.loading.value = false
    }
  }

  public async refreshUser () {
    this.loading.value = true
    try {
      const user = await this.cwaFetch.fetch('/me')
      this.authCookie.value = '1'
      this.authStore.data.user = user
      return user
    } catch (error) {
      if (!(error instanceof FetchError)) {
        throw error
      }
      await this.clearSession()
      return error
    } finally {
      this.loading.value = false
    }
  }

  public async init () {
    if (!this.hasCheckedMeEndpointForInit || (this.signedIn.value && !this.user)) {
      this.hasCheckedMeEndpointForInit = true
      await this.refreshUser()
    }
  }

  public get user () {
    return this.authStore.data.user
  }

  public get roles () {
    if (!this.user) {
      return
    }
    return this.user.roles
  }

  public get signedIn () {
    return computed(() => this.status.value === CwaAuthStatus.SIGNED_IN)
  }

  public hasRole (role: CwaUserRoles|string) {
    if (!this.roles) {
      return false
    }
    return this.roles.includes(role)
  }

  private async loginRequest (credentials: Credentials) {
    this.loading.value = true
    try {
      return await this.cwaFetch.fetch('/login', {
        method: 'POST',
        body: credentials
      })
    } catch (error) {
      this.loading.value = false
      if (!(error instanceof FetchError)) {
        throw error
      }
      return error
    }
  }

  public get status (): ComputedRef<CwaAuthStatus> {
    return computed(() => {
      if (this.loading.value) {
        return CwaAuthStatus.LOADING
      }

      return `${this.authCookie.value}` === '1' ? CwaAuthStatus.SIGNED_IN : CwaAuthStatus.SIGNED_OUT
    })
  }

  private async clearSession () {
    this.authStore.data.user = undefined
    this.authCookie.value = '0'
    this.admin.toggleEdit(false)
    this.mercure.init(true)
    this.resourcesStore.clearResources()
    this.fetcherStore.clearFetches()
    const route = useRoute()
    await this.fetcher.fetchRoute(route)
  }

  private get authStore () {
    return this.authStoreDefinition.useStore()
  }

  private get resourcesStore () {
    return this.resourcesStoreDefinition.useStore()
  }

  private get fetcherStore () {
    return this.fetcherStoreDefinition.useStore()
  }
}
