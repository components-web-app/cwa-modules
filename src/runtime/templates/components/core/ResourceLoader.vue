<template>
  <CwaUtilsAlertWarning v-if="!props.iri">
    <p>No IRI has been passed as a property to the `ResourceLoader` component</p>
  </CwaUtilsAlertWarning>
  <div v-else-if="isLoading">
    <CwaUtilsSpinner :show="true" />
  </div>
  <CwaUtilsAlertWarning v-else-if="(!resolvedComponent && !hasError) || (hasError && !hasSilentError)">
    <p v-if="!resource">
      Resource not found with iri `{{ props.iri }}`
    </p>
    <p v-else-if="!resource?.data">
      Resource not found with iri `{{ props.iri }}` data
    </p>
    <p v-else>
      The component `{{ resourceUiComponent }}` cannot be found
    </p>
  </CwaUtilsAlertWarning>
  <component v-bind="$attrs" :is="resolvedComponent" v-else-if="!hasError" :iri="props.iri" />
</template>

<script setup>
import { computed, onMounted, watch, getCurrentInstance, ref } from 'vue'
import { CwaResourceApiStatuses } from '../../../storage/stores/resources/state'
import { CwaAuthStatus } from '../../../api/auth'
import { useCwaResource, iri, useCwa } from '#imports'

const $cwa = useCwa()

const props = defineProps({
  ...iri,
  componentPrefix: {
    type: String,
    required: false,
    default: ''
  },
  uiComponent: {
    type: Object,
    required: false,
    default: undefined
  }
})

const resource = useCwaResource(props.iri)

const resourceLoadPendingStart = ref<Boolean>(!resource.value)
setTimeout(() => {
  if (resourceLoadPendingStart.value && resource.value) {
    resourceLoadPendingStart.value = false
  }
}, 20)

const isLoading = computed(() => {
  if (resourceLoadPendingStart.value) {
    return true
  }
  return !resource.value.data && resource.value.apiState.status === CwaResourceApiStatuses.IN_PROGRESS
})

const resourceUiComponent = computed(() => {
  if (!resource.value || !resource.value?.data) {
    return
  }
  return props.componentPrefix + (resource.value.data.uiComponent || resource.value.data['@type'])
})

const hasError = computed(() => {
  return resource.value?.apiState.status === CwaResourceApiStatuses.ERROR
})

const hasSilentError = computed(() => {
  return hasError.value && resource.value?.apiState.error.statusCode >= 400 && resource.value?.apiState.error.statusCode < 500
})

const resolvedComponent = computed(() => {
  if (props.uiComponent) {
    return props.uiComponent
  }

  const instance = getCurrentInstance()

  if (typeof instance?.appContext.components !== 'object' || !(resourceUiComponent.value in instance.appContext.components)) {
    return
  }

  return resourceUiComponent.value
})

onMounted(() => {
  watch(() => [$cwa.auth.status, hasSilentError, resource], async ([authStatus, hasSilentError, resource]) => {
    if (!resource.value?.data && hasSilentError.value && authStatus.value === CwaAuthStatus.SIGNED_IN) {
      await $cwa.fetchResource({
        path: props.iri
      })
    }
  }, {
    immediate: true
  })
})

</script>
