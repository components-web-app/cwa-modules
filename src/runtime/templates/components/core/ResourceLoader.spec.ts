// @vitest-environment nuxt
// todo: SOMETHING in this test file intermittently causes hanging in CI testing, async does not complete? All tests pass though...
import { beforeEach, describe, expect, test, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import { ref } from 'vue'
import * as vue from 'vue'
import ResourceLoader from './ResourceLoader.vue'
import { CwaAuthStatus } from '#cwa/runtime/api/auth'
import { CwaResourceApiStatuses } from '#cwa/runtime/storage/stores/resources/state'
import * as cwaComposables from '#cwa/runtime/composables/cwa'

const mockPrefix = 'TestComponent'
const mockIri = 'testIri'

vi.mock('vue', async () => {
  const mod = await vi.importActual<typeof import('vue')>('vue')
  return {
    ...mod,
    watch: vi.fn(() => {})
  }
})

function createWrapper (resource?: any, status?: CwaAuthStatus, component?: any) {
  // @ts-ignore
  vi.spyOn(cwaComposables, 'useCwa').mockImplementationOnce(() => ({
    auth: {
      status: {
        value: status ?? CwaAuthStatus.SIGNED_IN
      }
    },
    fetchResource: vi.fn(),
    resources: {
      getResource: vi.fn(() => ref(resource))
    }
  }))

  return mount(ResourceLoader, {
    shallow: true,
    props: {
      iri: mockIri,
      componentPrefix: mockPrefix,
      uiComponent: component
    },
    global: {
      renderStubDefaultSlot: true,
      components: {
        GlobalComponent: {
          name: 'GlobalComponent',
          template: '<div> global </div>',
          props: ['iri']
        }
      }
    }
  })
}

describe('ResourceLoader', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('reactive variables', () => {
    describe('resourceLoadBuffering', () => {
      test('If the resource exists, resourceLoadBuffering should be false', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.IN_PROGRESS
          }
        })
        expect(wrapper.vm.resourceLoadBuffering).toEqual(false)
      })

      test('If the resource does not exist, resourceLoadBuffering should be true and then revert to false after 20ms', () => {
        vi.useFakeTimers()
        const wrapper = createWrapper(undefined)
        expect(wrapper.vm.resourceLoadBuffering).toEqual(true)
        vi.advanceTimersByTime(20)
        expect(wrapper.vm.resourceLoadBuffering).toEqual(false)
        vi.useRealTimers()
      })
    })

    describe('isLoading', () => {
      test.each([
        {
          resourceLoadBuffering: false,
          resource: {
            data: null,
            apiState: {
              status: CwaResourceApiStatuses.IN_PROGRESS
            }
          },
          result: true
        },
        {
          resourceLoadBuffering: false,
          resource: {
            data: undefined,
            apiState: {
              status: CwaResourceApiStatuses.SUCCESS
            }
          },
          result: false
        },
        {
          resourceLoadBuffering: false,
          resource: {
            data: { test: true },
            apiState: {
              status: CwaResourceApiStatuses.IN_PROGRESS
            }
          },
          result: false
        },
        {
          resourceLoadBuffering: true,
          resource: {
            data: { test: true },
            apiState: {
              status: CwaResourceApiStatuses.IN_PROGRESS
            }
          },
          result: false
        },
        {
          resourceLoadBuffering: true,
          resource: undefined,
          result: true
        }
      ])('should return $result IF there resource is $resource and buffering is $resourceLoadBuffering', ({
        resource,
        result,
        resourceLoadBuffering
      }) => {
        const wrapper = createWrapper(resource)
        wrapper.vm.resourceLoadBuffering = ref(resourceLoadBuffering)
        expect(wrapper.vm.isLoading).toEqual(result)
      })
    })

    describe('resourceUiComponent', () => {
      test('should return nothing IF no resource is present', () => {
        const wrapper = createWrapper({ data: null, apiState: {} })

        expect(wrapper.vm.resourceUiComponent).not.toBeDefined()
      })

      test('should return nothing IF resource data is empty', () => {
        const wrapper = createWrapper({
          data: null,
          apiState: {
            status: CwaResourceApiStatuses.SUCCESS
          }
        })

        expect(wrapper.vm.resourceUiComponent).not.toBeDefined()
      })

      test('should return component name consisting of prefix and name from resource property', () => {
        const mockComponentName = 'Test'
        const wrapper = createWrapper({
          data: {
            uiComponent: mockComponentName
          },
          apiState: {
            status: CwaResourceApiStatuses.SUCCESS
          }
        })

        expect(wrapper.vm.resourceUiComponent).toEqual(mockPrefix + mockComponentName)
      })

      test('should return component name consisting of prefix and type from resource property', () => {
        const mockType = 'MType'
        const wrapper = createWrapper({
          data: {
            '@type': mockType
          },
          apiState: {
            status: CwaResourceApiStatuses.SUCCESS
          }
        })

        expect(wrapper.vm.resourceUiComponent).toEqual(mockPrefix + mockType)
      })
    })

    describe('hasError', () => {
      test('should return true IF api status EQUALS to error', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.ERROR,
            error: {}
          }
        })

        expect(wrapper.vm.hasError).toEqual(true)
      })

      test('should return false IF api status DOES NOT EQUAL to error', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.SUCCESS,
            error: {}
          }
        })

        expect(wrapper.vm.hasError).toEqual(false)
      })
    })

    describe('hasSilentError', () => {
      test('should return false IF hasError property is false', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.SUCCESS,
            error: {}
          }
        })

        expect(wrapper.vm.hasSilentError).toEqual(false)
      })

      test('should return false IF hasError property is true AND error status code is greater than 500', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.ERROR,
            error: {
              statusCode: 501
            }
          }
        })

        expect(wrapper.vm.hasSilentError).toEqual(false)
      })

      test('should return false IF hasError property is true AND error status code is less than 400', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.ERROR,
            error: {
              statusCode: 399
            }
          }
        })

        expect(wrapper.vm.hasSilentError).toEqual(false)
      })

      test('should return true IF hasError property is true AND error status code is within range of [400:500)', () => {
        const wrapper = createWrapper({
          apiState: {
            status: CwaResourceApiStatuses.ERROR,
            error: {
              statusCode: 401
            }
          }
        })

        expect(wrapper.vm.hasSilentError).toEqual(true)
      })
    })

    describe('resolvedComponent', () => {
      test('should return value from props IF it is defined', () => {
        const mockUiComponent = { name: 'MockComponent', template: '<div>I am mock</div>' }
        const wrapper = createWrapper({
          data: null,
          apiState: {
            status: CwaResourceApiStatuses.SUCCESS
          }
        }, CwaAuthStatus.SIGNED_IN, mockUiComponent)

        expect(wrapper.vm.resolvedComponent).toEqual(mockUiComponent)
      })
    })
  })

  describe('methods', () => {
    describe('fetchResource', () => {
      test('should NOT fetch resource IF ssr flag for api state is false', async () => {
        const resource = { apiState: { ssr: false } }
        const wrapper = createWrapper(resource)

        await wrapper.vm.methods.fetchResource()

        expect(wrapper.vm.$cwa.fetchResource).not.toHaveBeenCalled()
      })

      test('should NOT fetch resource IF ssr flag for api state is true AND resource has data', async () => {
        const resource = { apiState: { ssr: true }, data: { mock: true } }
        const wrapper = createWrapper(resource)

        await wrapper.vm.methods.fetchResource()

        expect(wrapper.vm.$cwa.fetchResource).not.toHaveBeenCalled()
      })

      test('should NOT fetch resource IF ssr flag for api state is true AND resource has no data AND no silent error', async () => {
        const resource = { apiState: { ssr: true, status: CwaResourceApiStatuses.SUCCESS }, data: null }
        const wrapper = createWrapper(resource)

        await wrapper.vm.methods.fetchResource()

        expect(wrapper.vm.$cwa.fetchResource).not.toHaveBeenCalled()
      })

      test('should fetch resource IF resource data is empty, silent error occurred, resource was fetched during SSR', async () => {
        const resource = { apiState: { ssr: true, error: { statusCode: 404 }, status: CwaResourceApiStatuses.ERROR }, data: undefined }
        const wrapper = createWrapper(resource)

        await wrapper.vm.methods.fetchResource()

        expect(wrapper.vm.$cwa.fetchResource).toHaveBeenCalledWith({ path: mockIri })
      })
    })
  })

  describe('watch', () => {
    test('should be called with correct callback and options', () => {
      const watchSpy = vue.watch
      const wrapper = createWrapper({
        apiState: {
          status: CwaResourceApiStatuses.IN_PROGRESS
        }
      })

      expect(watchSpy.mock.calls[0][0]).toHaveLength(2)
      expect(watchSpy.mock.calls[0][0][0].value).toEqual(wrapper.vm.hasSilentError)
      expect(watchSpy.mock.calls[0][0][1].value).toEqual(wrapper.vm.resource)
      // expect(watchSpy.mock.calls[0][0]).toEqual([wrapper.vm.hasSilentError, wrapper.vm.resource])
      expect(watchSpy.mock.calls[0][1]).toEqual(wrapper.vm.methods.fetchResource)
      expect(watchSpy.mock.calls[0][2]).toEqual({ immediate: true })
    })
  })

  describe('snapshots', () => {
    test('should match snapshot IF resource is loading', () => {
      const wrapper = createWrapper({ data: null, apiState: { status: CwaResourceApiStatuses.IN_PROGRESS } })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF resource does not exist after loading', async () => {
      vi.useFakeTimers()
      const wrapper = createWrapper(undefined)
      await vi.advanceTimersByTimeAsync(20)
      expect(wrapper.element).toMatchSnapshot()
      vi.useRealTimers()
    })

    test('should match snapshot IF resource data is not resolved', () => {
      const wrapper = createWrapper({
        data: null,
        apiState: {
          status: CwaResourceApiStatuses.SUCCESS
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF UI component is not resolved', () => {
      const wrapper = createWrapper({
        data: {
          uiComponent: 'Mock'
        },
        apiState: {
          status: CwaResourceApiStatuses.ERROR,
          error: {}
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF component is resolved from props', async () => {
      const wrapper = createWrapper({
        data: {
          uiComponent: 'DummyComponent'
        },
        apiState: {
          status: CwaResourceApiStatuses.SUCCESS
        }
      })

      await wrapper.setProps({
        uiComponent: {
          name: 'DummyComponent',
          template: '<div> test </div>',
          props: ['iri']
        }
      })

      expect(wrapper.element).toMatchSnapshot()
    })

    test('should match snapshot IF global component name is resolved', async () => {
      const wrapper = createWrapper({
        data: {
          uiComponent: 'GlobalComponent'
        },
        apiState: {
          status: CwaResourceApiStatuses.SUCCESS
        }
      })
      await wrapper.setProps({
        componentPrefix: undefined
      })

      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
