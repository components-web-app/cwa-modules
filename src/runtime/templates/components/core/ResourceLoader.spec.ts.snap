// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResourceLoader > snapshots > should match snapshot IF UI component is not resolved 1`] = `
<warning-stub>
  <p>
     The component \`TestComponentMock\` for resource \`testIri\` cannot be resolved 
  </p>
</warning-stub>
`;

exports[`ResourceLoader > snapshots > should match snapshot IF component is resolved from props 1`] = `
<dummycomponent-stub
  iri="testIri"
/>
`;

exports[`ResourceLoader > snapshots > should match snapshot IF global component name is resolved 1`] = `
<globalcomponent-stub
  iri="testIri"
/>
`;

exports[`ResourceLoader > snapshots > should match snapshot IF resource data is not resolved 1`] = `
<warning-stub>
  <p>
     No data received for resource \`testIri\` 
  </p>
</warning-stub>
`;

exports[`ResourceLoader > snapshots > should match snapshot IF resource does not exist after loading 1`] = `
<warning-stub>
  <p>
     Resource \`testIri\` has not been requested 
  </p>
</warning-stub>
`;

exports[`ResourceLoader > snapshots > should match snapshot IF resource is loading 1`] = `
<div>
  <spinner-stub
    show="true"
  />
</div>
`;
